### PR TITLE
Spec deprecatedURNtoURL

### PR DESCRIPTION
**Pull request by JensenPaul**
Addresses #160

***
<a href="https://pr-preview.s3.amazonaws.com/JensenPaul/fenced-frame/pull/161.html" title="Last updated on Jun 6, 2024, 1:33 PM UTC (6495120)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/fenced-frame/161/4966a9e...JensenPaul:6495120.html" title="Last updated on Jun 6, 2024, 1:33 PM UTC (6495120)">Diff</a>

### CODE REVIEW

File: spec.bs [hunk 1]
```diff
@@ -1437,12 +1437,20 @@ the API {{Window/navigator}}.{{Navigator/deprecatedReplaceInURN()}}, which allow
 macros into the [=fenced frame config/mapped url=] corresponding to a given [=urn uuid=] or
 {{FencedFrameConfig}}.
 
+Note: To help with ease of adoption,
```
Comment (qingxinwu) [inline spec.bs:1440]: Does this one have a different plan from deprecatedReplaceInURN()? The note above says until 2026, and here it says third party cookie deprecation. If they have the same plan, we should probably update both, or possibly a single note paragraph to cover both the two APIs?

Comment (JensenPaul) [reply spec.bs:1440]: They're different timelines, hence the different wording and different links.

**Review by qingxinwu — APPROVED** [review]

**Comment by qingxinwu** [comment]
LGTM

**Comment by domfarolino** [comment]
I saw this in my inbox but I noticed I was not added as a reviewer. Should I take a look? Feel free to add me as a reviewer and I will take a look after I'm back from vacation.

File: spec.bs [hunk 2]
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
...
+  1. Let |p| be [=a new promise=].
+
+  1. Let |global| be [=this=]'s [=relevant global object=].
+
+  1. Run the following steps [=in parallel=]:
+
+     1. Let |mapping| be |global|'s [=associated Document=]'s [=node navigable=]'s
+        [=navigable/traversable navigable=]'s [=traversable navigable/fenced frame config mapping=].
+
+     1. Let |config| be the result of [=fenced frame config mapping/finding a config=] in |mapping|
```
[the remark this answers is not in the document]
Comment (JensenPaul) [reply spec.bs:1537]: Does that mean callers have to wait for a Protected Audience auction to complete before calling `deprecatedURNtoURL()`?
It does look like we require the mapping to be finalized: https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/ad_auction_service_impl.cc;l=487;drc=2b8b8c15ea96feadd34e04ca947e5f05069e3b24
Should I just change this to instead say this?
```
    1. If |mapping|'s [=fenced frame config mapping/finalized config mapping=][|urn|], does not
       [=map/exist=], [=queue a global task=] on the [=DOM manipulation task source=]
       given |global|, to [=resolve=] |p| with {{undefined}}, and abort these steps.
    1. Let |config| be |mapping|'s [=fenced frame config mapping/finalized config mapping=][|urn|].
```

[the remark this answers is not in the document]
Comment (gtanzer) [reply spec.bs:1537]: It's not an issue for PA because the pending urn is never returned to script, so it can't be passed to deprecatedURNtoTURL; instead the promise waits for the auction result to be resolved. In selectURL it immediately resolves to a pending urn, so you could pass it as input (and it won't work).

[the remark this answers is not in the document]
Comment (JensenPaul) [reply spec.bs:1537]: So what's the right thing to put in the spec?  it sounds like it only works with finalized URNs, in which case I think my suggestion in my last comment is accurate?

[the remark this answers is not in the document]
Comment (gtanzer) [reply spec.bs:1537]: Yeah, afaict what you put in the comment matches the implementation.

[the remark this answers is not in the document]
Comment (JensenPaul) [reply spec.bs]: Okay, I switched to my suggested steps.

File: spec.bs [hunk 3]
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
...
+        [=map/exist=], [=queue a global task=] on the [=DOM manipulation task source=]
+        given |global|, to [=resolve=] |p| with {{undefined}}, and abort these steps.
+
+     1. Let |config| be |mapping|'s [=fenced frame config mapping/finalized config mapping=][|urn|].
+
+     1. [=Queue a global task=] on the [=DOM manipulation task source=] given |global|, to
+        [=resolve=] |p| with |config|'s [=fenced frame config/mapped url=]'s [=mapped url/value=].
+
+     1. If |send_reports| is true, run steps in |config|'s
+        [=fenced frame config/on navigate callback=].
```
Comment (domfarolino) [inline spec.bs:1547]: Why run these just on lookup? I assume that's what the imple does? It just feels a little strange.

**Review by domfarolino — COMMENTED** [review]
Pretty much lgtm; nothing major here.

File: spec.bs (hunk 3 above)
Comment (JensenPaul) [reply spec.bs:1547]: `deprecatedURNtoURL()` was added eons ago and the motivation was stated in https://github.com/WICG/turtledove/blob/main/Proposed_First_FLEDGE_OT_Details.md#advertisement-rendering as: "the `sendReports` parameter...is included to [promote accurate measurement](https://github.com/WICG/turtledove/blob/main/Proposed_First_FLEDGE_OT_Details.md#promoting-accurate-measurement-of-fledge-effectiveness.) and comparison of FLEDGE API utility by allowing auction winning URLs to be treated and used as they are in today’s ad auctions."
Basically we wanted a way to test Protected Audience (especially reporting scripts and APIs) when the URL coming from `deprecatedURNtoURL()` was rendered in a some other way that's not visible to the browser (so the browser doesn't know when to trigger Protected Audience reporting without the hint from `sendReports`).

File: spec.bs [hunk 4]
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
...
+
+  1. If |urn| is <span class=XXX>TODO invalid</span>, [=exception/throw=] a {{TypeError}}.
+
+  1. Let |p| be [=a new promise=].
+
+  1. Let |global| be [=this=]'s [=relevant global object=].
+
+  1. Run the following steps [=in parallel=]:
+
+     1. Let |mapping| be |global|'s [=associated Document=]'s [=node navigable=]'s
```
[the remark this answers is not in the document]
Comment (JensenPaul) [reply spec.bs]: I think I fixed this.  Please take a look.

File: spec.bs [hunk 5]
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
   </wpt>
 </div>
 
+<div algorithm>
+  The <dfn method for=Navigator>deprecatedURNtoURL(|urnOrConfig|, |send_reports|)</dfn>
+  method steps are:
```
Comment (domfarolino) [inline spec.bs:1518]: All of this can be wrapped like so:

```
  The <dfn method for=Navigator>deprecatedURNtoURL(|urnOrConfig|, |send_reports|)</dfn> method
  steps are:
```

and be compliant with 100 lines.

File: spec.bs (hunk 3 above)
Comment (domfarolino) [reply spec.bs:1547]: Any chance there's a WPT for this?

File: spec.bs [hunk 6]
```diff
@@ -1505,6 +1513,46 @@ partial interface Navigator {
...
+
+  1. Let |urn| be null.
+
+  1. If |urnOrConfig| is a {{USVString}}, set |urn| to |urnOrConfig|.
+
+  1. Otherwise, set |urn| to |urnOrConfig|'s [=fencedframeconfig/urn=].
+
+  1. If |urn| is <span class=XXX>TODO invalid</span>, then [=exception/throw=] a {{TypeError}}.
```
Comment (domfarolino) [inline spec.bs:1526]: https://github.com/WICG/fenced-frame/pull/161#discussion_r1626061123 got marked outdated, so I guess I'll re-file this comment as a stand-in for it.

**Review by domfarolino — COMMENTED** [review]
Btw what is the motivation to include this API in this spec, instead of the PA spec? Impl-wise, it looks like it was included closer to the PA stuff: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ad_auction/navigator_auction.idl;l=41;drc=ac83a5a2d3c04763d86ce16d92f3904cc9566d3a;bpv=1;bpt=0

File: spec.bs (hunk 6 above)
Comment (JensenPaul) [reply spec.bs]: I proposed some new words here, WDYT?

File: spec.bs (hunk 3 above)
Comment (JensenPaul) [reply spec.bs:1549]: It looks like we don't have WPTs for when `sendReports` is `true`, but I should note that `deprecatedURNtoURL()` is [going away at 3PCD](https://developers.google.com/privacy-sandbox/relevance/protected-audience-api/feature-status#:~:text=Note%3A%20The%20temporary%20navigator.deprecatedURNToURL()%20will%20be%20removed%20by%20third%2Dparty%20cookie%20deprecation.).

**Comment by JensenPaul** [comment]
> Btw what is the motivation to include this API in this spec, instead of the PA spec? Impl-wise, it looks like it was included closer to the PA stuff: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ad_auction/navigator_auction.idl;l=41;drc=ac83a5a2d3c04763d86ce16d92f3904cc9566d3a;bpv=1;bpt=0

I feel like stuff that only relates to URNs, like `deprecatedURNtoURL()`, should be in the FF spec, like `deprecatedReplaceInURN()` already is.  The use of URNs (and then FFConfig) seems tied to seeding FFs, and could be used by specs other than PA.

**Comment by domfarolino** [comment]
Ah darn it looks there are conflicts now that the other PR merged. Would you be able to sort them out please?

**Comment by JensenPaul** [comment]
> Ah darn it looks there are conflicts now that the other PR merged. Would you be able to sort them out please?

I fixed the merge conflict.  Now it looks like the build is failing on an unrelated line:
`LINE 1724: Couldn't find WPT test 'fenced-frame/fence-report-event-cross-origin.https.html' - did you misspell something?`
Looks related to https://chromium-review.googlesource.com/c/chromium/src/+/5544804
So I updated the WPT file name.

File: spec.bs [hunk 7]
```diff
@@ -1507,7 +1516,47 @@ partial interface Navigator {
...
+  The <dfn method for=Navigator>deprecatedURNtoURL(|urnOrConfig|, |send_reports|)</dfn> method
+  steps are:
+
+  1. Let |urn| be null.
+
+  1. If |urnOrConfig| is a {{USVString}}, set |urn| to |urnOrConfig|.
+
+  1. Otherwise, set |urn| to |urnOrConfig|'s [=fencedframeconfig/urn=].
 
+  1. If |urn| is not a valid [=urn uuid=] (i.e. won't pass the ABNF in Section 3 of
```
Comment (domfarolino) [inline spec.bs:1528]:
```suggestion
  1. If |urn| is not a valid [=urn uuid=] (i.e., won't pass the ABNF in Section 3 of
```

nit: https://whatwg.org/style-guide#grammar

File: spec.bs (hunk 6 above)
Comment (domfarolino) [reply spec.bs]: I think that sounds right? Although it would be inelegant to have to write that `i.e.,` each time we reference the concept of an invalid URN. But I suppose this is an incremental improvement so that's fine.

**Review by domfarolino — APPROVED** [review]
(Meant to "approve" with nits, whoops)

**Review by NameNoQuality — APPROVED** [review]
